server/ServerFSM.cpp

```diff
@@ -2324,6 +2324,20 @@ sc::result ProcessingTurn::react(const ProcessTurn& u) {
         }
     }
 
+    if (server.IsHostless() && GetOptionsDB().Get<bool>("save.auto.hostless.enabled")) {
```

Comment:
I think autosaves are supposed to be done at the end of a turn, not at the start.

Reply:
Doesn't turn end in the end of `ProcessingTurn::react(const ProcessTurn&)`?

Reply:
I mean after all players have given orders, it should autosave, then process those orders and the between-turn events and updates. If the save is loaded, players should be able to issue more orders and re-end the turn. At least that's how it should work in single player.

Reply:
I'll try to change it but I wonder which correct state and event sequence to save a game now.

Reply:
I don't see how the game saved after getting all orders and before turn processing. We waits for orders in `TurnOrder` and raise `CheckTurnEndConditions` where the server goes to `ProcessingTurn` state when all orders are received.

Reply:
Singleplayer autosaves are initiated by the player's client when the player clicks end turn, before the actual end turn message is sent, in `HumanClientApp::StartTurn`

Reply:
The client sends `SAVE_GAME_INITIATE` which translates to `SaveGameRequest` event, but only `WaitingForTurnEndIdle` is a state processes `SaveGameRequest` where the server goes from `WaitingForSaveData` after game saved or from `ProcessingTurn` after `ProcessTurn` has been handled.

I don't see a way how the game can save something between acquiring orders and turn processing.

Reply:
When the client sends the request, the server is still in `WaitingForTurnEndIdle`.

Reply:
Yes, but it before all players sent their orders.

Reply:
Yes, but other players can send their orders after host.

Reply:
not in single player.

in multiplayer, I think the autosaves are done at the start of a turn. so I guess this timing is OK.

Reply:
Yes, after all orders have been received and processed there will be start of next turn where I push `SaveGameRequest` for hostless autosave.